identity: Fix a minor leak using meta_str

diff --git a/plugins/elements/gstidentity.c b/plugins/elements/gstidentity.c
index 8da7cf068..e97fe498e 100644
--- a/plugins/elements/gstidentity.c
+++ b/plugins/elements/gstidentity.c
@@ -612,6 +612,7 @@ gst_identity_update_last_message_for_buffer (GstIdentity * identity,
       GST_BUFFER_OFFSET (buf), GST_BUFFER_OFFSET_END (buf),
       GST_BUFFER_FLAGS (buf), flag_str, meta_str ? meta_str : "none", buf);
   g_free (flag_str);
+  g_free (meta_str);
 
   GST_OBJECT_UNLOCK (identity);
 