diff -Naur before/omx/gstomxmjpegdec.c 1.16.2-r0/gst-omx-1.16.2/omx/gstomxmjpegdec.c
--- before/omx/gstomxmjpegdec.c	2020-10-12 18:00:56.336421374 +0900
+++ 1.16.2-r0/gst-omx-1.16.2/omx/gstomxmjpegdec.c	2020-10-12 18:26:37.394386744 +0900
@@ -59,7 +59,7 @@
       GST_DEBUG_FUNCPTR (gst_omx_mjpeg_dec_is_format_change);
   videodec_class->set_format = GST_DEBUG_FUNCPTR (gst_omx_mjpeg_dec_set_format);
 
-  videodec_class->cdata.default_sink_template_caps = "image/jpeg, "
+  videodec_class->cdata.default_sink_template_caps = "video/x-jpeg, "
       "width=(int) [1,MAX], " "height=(int) [1,MAX]";
 
   gst_element_class_set_static_metadata (element_class,
