--- 1.16.2-r0/gst-omx-1.16.2/omx/gstomxaudiodec.c	2021-11-22 14:20:30.073934667 +0900
+++ gst-omx-1.16.2/omx/gstomxaudiodec.c	2021-11-22 14:20:45.485950343 +0900
@@ -715,6 +715,8 @@
       gst_pad_pause_task (GST_AUDIO_DECODER_SRC_PAD (self));
       self->started = FALSE;
       g_mutex_unlock (&self->drain_lock);
+    } else if(flow_ret == GST_FLOW_NOT_LINKED){
+      self->downstream_flow_ret = GST_FLOW_OK;
     }
     GST_AUDIO_DECODER_STREAM_UNLOCK (self);
     return;
